### PR TITLE
Observable collection Move behavior #355

### LINKED ISSUE #355: Wrong ObservableCollection EventArg after  drag/drop
I'm using ObservableCollection<T> which is bound to ItemsControl ItemsSource property. 

What I expect to get after drag and drop an item within control is NotifyCollectionChangedAction.Move event like this:

```
        private void MyCollection_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
        {
            if (e.Action == NotifyCollectionChangedAction.Move)
            {
                //DO something
            }
        }
```

But instead I'm getting two events (NotifyCollectionChangedAction.Remove and NotifyCollectionChangedAction.Add). I'm not sure if this is by design but in my case it was complicating the logic to decide if there was new item added externally or just drag/dropped existing control item.

### Repo
I can provide code example if needed.
### Environment

- GongSolutions.WPF.DragDrop 2.2.0
- Windows 10
- Visual Studio 2019
- .NET Framework 4.7.2

### PR DESCRIPTION
**Pull request by Mogikan**
## What changed?

Instead of delete/insert move is called.

Closes #355

### CODE REVIEW

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 1]
```diff
@@ -181,23 +183,27 @@ public virtual void Drop(IDropInfo dropInfo)
 
             var destinationList = dropInfo.TargetCollection.TryGetList();
             var data = ExtractData(dropInfo.Data).OfType<object>().ToList();
-
+            bool forceMoveBehavior = false;
             var copyData = ShouldCopyData(dropInfo);
             if (!copyData)
             {
-                var sourceList = dropInfo.DragInfo.SourceCollection.TryGetList();
+                var sourceList = dropInfo.DragInfo.SourceCollection.TryGetList();
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:190]: Should avoid adding whitespace to the end of lines.
```suggestion
                var sourceList = dropInfo.DragInfo.SourceCollection.TryGetList();
```

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 2]
```diff
@@ -220,10 +226,22 @@ public virtual void Drop(IDropInfo dropInfo)
                         {
                             obj2Insert = cloneable.Clone();
                         }
-                    }
+                    }
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:229]: Should avoid adding whitespace to the end of lines.
```suggestion
                    }
```

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 3]
```diff
@@ -234,6 +252,30 @@ public virtual void Drop(IDropInfo dropInfo)
...
+        private static void Move(IList list, int sourceIndex, int destinationIndex)
+        {
+            if (!IsObservableCollection(list))
+                throw new ArgumentException("ObservableCollection<T> was expected",nameof(list));
+            var method = list.GetType().GetMethod("Move",
+                System.Reflection.BindingFlags.Instance
+                | System.Reflection.BindingFlags.Public);            
+            //method = method.MakeGenericMethod(listItemType);
+            method.Invoke(list, new object[] { sourceIndex, destinationIndex });            
+
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:264]: Unnecessary blank line
```suggestion
```

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 4]
```diff
@@ -234,6 +252,30 @@ public virtual void Drop(IDropInfo dropInfo)
...
         }
 
+        private static void Move(IList list, int sourceIndex, int destinationIndex)
+        {
+            if (!IsObservableCollection(list))
+                throw new ArgumentException("ObservableCollection<T> was expected",nameof(list));
+            var method = list.GetType().GetMethod("Move",
+                System.Reflection.BindingFlags.Instance
+                | System.Reflection.BindingFlags.Public);            
+            //method = method.MakeGenericMethod(listItemType);
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:262]: Should clean up comment out code
```suggestion
```

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 5]
```diff
@@ -234,6 +252,30 @@ public virtual void Drop(IDropInfo dropInfo)
             }
         }
 
+        private static void Move(IList list, int sourceIndex, int destinationIndex)
+        {
+            if (!IsObservableCollection(list))
+                throw new ArgumentException("ObservableCollection<T> was expected",nameof(list));
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:258]: Should have a space between parameters.
```suggestion
                throw new ArgumentException("ObservableCollection<T> was expected", nameof(list));
```

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 6]
```diff
@@ -1,8 +1,10 @@
 ﻿using System;
 using System.Collections;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
 using System.ComponentModel;
 using System.Linq;
+using System.Reflection;
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:7]: We don't need to add this `using` as both uses of `BindingFlags` below has the namespace included.
```suggestion
```

File: src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs [hunk 7]
```diff
@@ -234,6 +252,30 @@ public virtual void Drop(IDropInfo dropInfo)
...
+                System.Reflection.BindingFlags.Instance
+                | System.Reflection.BindingFlags.Public);            
+            //method = method.MakeGenericMethod(listItemType);
+            method.Invoke(list, new object[] { sourceIndex, destinationIndex });            
+
+        }
+
+        private static bool SameObservableCollection(IList collection1, IList collection2)
+        {
+            return Equals(collection1, collection2) && IsObservableCollection(collection1);
```
Comment (jizc) [inline src/GongSolutions.WPF.DragDrop/DefaultDropHandler.cs:269]: Should use `ReferenceEquals` in this case.
```suggestion
            return ReferenceEquals(collection1, collection2) && IsObservableCollection(collection1);
```

**Review by jizc — CHANGES_REQUESTED** [review]
Here are my unsolicited suggestions 😉

**Review by Mogikan — COMMENTED** [review]
Thanks

**Comment by punker76** [comment]
@Mogikan your changes have been merged, thanks for your contribution 👍

**Pull request merged**